content/issue-2/applicative-parsing.md

```diff
@@ -0,0 +1,675 @@
+    fn fail(msg: &str) -> Self {
+        Magic(Err(msg.to_owned()))
+    }
+}
+```
+
+Customizing error messages:
+
+```rust
+let name = short.alt(long).alt(fail("You need to pass --long or --short"));
+```
+
+Fallback to some default value:
+
+```rust
+let name = short.alt(pure("Anonymous"));
+```
+
+## Defined operations summary
+
+Operations defined so far are sufficient to express arbitrary computations on command line
+arguments:
+
+```rust
+struct Magic<T>(Result<T, String>);
+impl<T> Magic<T> {
+    /// a way to run a computation
+    fn run(self) -> Result<T, String> {
+        self.0
+    }
+
+    /// apply a pure transformation to a contained value
+    fn map<R, F: Fn(T) -> R>(self, f: F) -> Magic<R> {
+        self.parse(|x| Ok(f(x)))
+    }
+
+    /// apply a failing transformation to a contained value
+    fn parse<R, F: Fn(T) -> Result<R, String>>(self, f: F) -> Magic<R> {
+        match self.0 {
+            Ok(x) => Magic(f(x)),
+            Err(err) => Magic(Err(err)),
+        }
+    }
+
+    /// compose two computations into a single one requiring both to succeed
+    fn zip<R>(self, other: Magic<R>) -> Magic<(T, R)> {
+        match (self.0, other.0) {
+            (Ok(t), Ok(r)) => Magic(Ok((t, r))),
+            (Err(err), _) | (_, Err(err)) => Magic(Err(err)),
+        }
+    }
+
+    /// compose two computations, picking succeeding one
+    fn alt(self, other: Self) -> Self {
+        match self.0 {
+            Ok(t) => Magic(Ok(t)),
+            Err(_) => other,
+        }
+    }
+
+    /// create an unconditionally succeeding computation
+    fn pure(val: T) -> Self {
+        Magic(Ok(val))
+    }
+
+    /// create an unconditionally failing computation
+    fn fail(msg: &str) -> Self {
+        Magic(Err(msg.to_owned()))
+    }
+}
+```
+
+Those seven operations serve as a base for the parser. They allow to compose primitive argument
+parsers in many ways to create a very wide range of computations and there's no mutations in the
+API itself so it fits perfectly with a functional programming style.
+
+There's a few examples of useful operations implemented in terms of this base API
+
+Optional command line arguments:
+
+```rust
+fn optional(magic: Magic<T>) -> Magic<Option<T>> {
+    // if magic contains a value - wrap it in Some, otherwise use None
+    // this will also consume any errors that can be inside, but we'll address
+    // this problem later
+    magic.map(Some).alt(pure(None))
+}
+```
+
+Validating values with a function:
+
+```rust
+fn guard(magic: Magic<T>, check: impl Fn(&T) -> bool, msg: &str) -> Magic<T> {
+    magic.parse(|val| {
+        if check(&val) {
+            Ok(val(
+        } else {
+            Err(msg.into())
+        }
+    })
+}
+```
+
+# Back to practical implementation
+
+Now that the parser has all the basic building blocks the next step is to reimplement them
+without `Magic<T>` since current version isn't practical.
```

Comment:
Hmm... Combinators are practical, representation is not. Something like this maybe?

```suggestion
Now that the parser has all the basic building blocks the next step is to reimplement them
without `Magic<T>` since current internal representation relies on magic to get from arguments on a command line to variables.
```